### PR TITLE
Fix mod extended builtin

### PR DESCRIPTION
**Pull request by jackvreeken**
- [x] Merge custom unit fix first

### CODE REVIEW

**Comment by jbaayen** [comment]
With this one I get:

```
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/rtctools/optimization/modelica_mixin.py", line 48, in __init__
    self.__pymoca_model = pymoca.backends.casadi.api.transfer_model(kwargs['model_folder'], model_name, self.compiler_options())
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/backends/casadi/api.py", line 302, in transfer_model
    model = _compile_model(model_folder, model_name, compiler_options)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/backends/casadi/api.py", line 94, in _compile_model
    model = generator.generate(tree, model_name, compiler_options)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/backends/casadi/generator.py", line 759, in generate
    ast_walker.walk(casadi_gen, flat_tree)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 176, in walk
    self.handle_walk(listener, tree.__dict__[child_name])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 193, in handle_walk
    self.handle_walk(listener, tree[k])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 190, in handle_walk
    self.walk(listener, tree)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 176, in walk
    self.handle_walk(listener, tree.__dict__[child_name])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 196, in handle_walk
    self.handle_walk(listener, tree[i])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 190, in handle_walk
    self.walk(listener, tree)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 176, in walk
    self.handle_walk(listener, tree.__dict__[child_name])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 196, in handle_walk
    self.handle_walk(listener, tree[i])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 190, in handle_walk
    self.walk(listener, tree)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 176, in walk
    self.handle_walk(listener, tree.__dict__[child_name])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 190, in handle_walk
    self.walk(listener, tree)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 176, in walk
    self.handle_walk(listener, tree.__dict__[child_name])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 196, in handle_walk
    self.handle_walk(listener, tree[i])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 190, in handle_walk
    self.walk(listener, tree)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 176, in walk
    self.handle_walk(listener, tree.__dict__[child_name])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 196, in handle_walk
    self.handle_walk(listener, tree[i])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 190, in handle_walk
    self.walk(listener, tree)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 176, in walk
    self.handle_walk(listener, tree.__dict__[child_name])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 196, in handle_walk
    self.handle_walk(listener, tree[i])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 190, in handle_walk
    self.walk(listener, tree)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 176, in walk
    self.handle_walk(listener, tree.__dict__[child_name])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 196, in handle_walk
    self.handle_walk(listener, tree[i])
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 190, in handle_walk
    self.walk(listener, tree)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/tree.py", line 180, in walk
    getattr(listener, 'exit' + name)(tree)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/backends/casadi/generator.py", line 263, in exitExpression
    rhs = ca.MX(self.get_mx(tree.operands[1]))
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/backends/casadi/generator.py", line 700, in get_mx
    s = self.get_component(tree)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/backends/casadi/generator.py", line 685, in get_component
    s = self.get_indexed_symbol(tree, s)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pymoca/backends/casadi/generator.py", line 652, in get_indexed_symbol
    assert s.size2() > 0, "Second dimension of matrix is zero"
AssertionError: Second dimension of matrix is zero
```

Model attached.

[Example.mo.txt](https://github.com/pymoca/pymoca/files/1843117/Example.mo.txt)

**Comment by vdwees** [comment]
Interesting. I guess this is because C is empty- no substances. I guess we should be dropping the equation, but I don't know where that should be handled.

**Comment by jackvreeken** [comment]
I think this fixes it. Zero-sized symbols seem to work correctly automatically in most places, with e.g. addition or multiplication resulting in another zero-sized symbol. The only issue was with the for loops trying to access the .name() property of these symbols, which is not possible.

Note that the map/vcat in exitForEquation also seems to work/do the right thing, where the zero sized equation just disappears.

File: src/pymoca/tree.py [hunk 1]
```diff
@@ -377,18 +388,38 @@ def build_instance_tree(orig_class: Union[ast.Class, ast.InstanceClass], modific
...
+            inheriting_from_builtin = extends_builtin(c)
+            sym_mod = ast.ClassModification()
             for arg in sym_arguments:
                 if arg.value.component.indices:
                     raise Exception("Subscripting modifiers is not allowed.")
 
-                arg.value.component = arg.value.component.child[0]
+                if inheriting_from_builtin:
+                    for el_arg in arg.value.modifications:
+                        if not isinstance(el_arg, ast.ClassModification):
```
Comment (jbaayen) [inline src/pymoca/tree.py]: Flip if/else blocks for clarity?

File: src/pymoca/tree.py [hunk 2]
```diff
@@ -377,18 +388,38 @@ def build_instance_tree(orig_class: Union[ast.Class, ast.InstanceClass], modific
...
+                            # If the value is being set, we make a new class
+                            # modification with attribute name "value" that we
+                            # pick up later in modify_symbol()
+
+                            # TODO: Figure out if it's easier to directly do this
+                            # in the parser.
```
Comment (jbaayen) [inline src/pymoca/tree.py]: Yeap that seems like the cleaner solution in the long run. Though I'd suggest we merge this for now, the TODO comment is in there after all.

File: src/pymoca/tree.py (hunk 1 above)
Comment (jackvreeken) [reply src/pymoca/tree.py]: Although I agree that checking for "not <something>" is a bit odd, I'd rather keep it the same as the equivalent check and block a couple lines up.

**Pull request merged**